server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -739,6 +753,50 @@ private void validateIndexTemplateV2(String name, ComposableIndexTemplate indexT
         }
     }
 
+    private void validateUseOfDeprecatedComponentTemplates(
+        String name,
+        ComposableIndexTemplate template,
+        Map<String, ComponentTemplate> componentTemplates
+    ) {
+        template.composedOf()
+            .stream()
+            .map(ct -> Tuple.tuple(ct, componentTemplates.get(ct)))
+            .filter(ct -> Objects.nonNull(ct.v2()))
+            .filter(ct -> ct.v2().deprecated())
+            .forEach(
+                ct -> deprecationLogger.warn(
```

Comment:
@dakrone what's a good way to test deprecation logging? Yaml rest tests or something else?

Reply:
You can test it either in yaml tests:

(look for tests with something like:)

```
warnings:
   - "Deprecated field [template] used, replaced by [index_patterns]"
```

Or else you can test it in a Java integ test with the `assertWarnings(…)` assertion.